pymbolic/mapper/equalizer.py

```diff
@@ -0,0 +1,169 @@
+                and self.rec(expr.right, other.right, *args, **kwargs)
+
+    def map_logical_not(self, expr, other, *args, **kwargs):
+        return type(expr) == type(other) \
+                and self.rec(expr.child, other.child, *args, **kwargs)
+
+    def map_logical_or(self, expr, other, *args, **kwargs):
+        return self._map_multichild_expr(expr, other, *args, **kwargs)
+
+    def map_logical_and(self, expr, other, *args, **kwargs):
+        return self._map_multichild_expr(expr, other, *args, **kwargs)
+
+    def map_list(self, expr, other, *args, **kwargs):
+        return expr == other
+
+    def map_tuple(self, expr, other, *args, **kwargs):
+        return expr == other
+
+    def map_numpy_array(self, expr, other, *args, **kwargs):
+        import numpy
+        return numpy.array_equal(expr, other)
+
+    def map_multivector(self, expr, other, *args, **kwargs):
+        return type(expr) == type(other) \
+                and self.data == other.data \
+                and self.space == other.space
+
+    def map_common_subexpression(self, expr, other, *args, **kwargs):
+        return type(expr) == type(other) \
+                and self.rec(expr.child, other.child, *args, **kwargs) \
+                and self.rec(expr.prefix, other.prefix, *args, **kwargs) \
+                and self.rec(expr.scope, other.scope, *args, **kwargs)
+
+    def map_if(self, expr, other, *args, **kwargs):
+        return type(expr) == type(other) \
+                and self.rec(expr.condition, other.condition, *args, **kwargs) \
+                and self.rec(expr.then, other.then, *args, **kwargs) \
+                and self.rec(expr.else_, other.else_, *args, **kwargs)
+
+    def map_if_positive(self, expr, other, *args, **kwargs):
+        return type(expr) == type(other) \
+                and self.rec(expr.criterion, other.criterion, *args, **kwargs) \
+                and self.rec(expr.then, other.then, *args, **kwargs) \
+                and self.rec(expr.else_, other.else_, *args, **kwargs)
+
+    def map_min(self, expr, other, *args, **kwargs):
+        return type(expr) == type(other) \
+                and self.rec(expr.children, other.children, *args, **kwargs)
+
+    map_max = map_min
+
+    def map_derivative(self, expr, other, *args, **kwargs):
+        return type(expr) == type(other) \
+                and self.rec(expr.child, other.child, *args, **kwargs) \
+                and self.rec(expr.varaibles, other.variables, *args, **kwargs)
+
+    def map_substitution(self, expr, other, *args, **kwargs):
+        return type(expr) == type(other) \
+               and self.rec(expr.child, other.child, *args, **kwargs) \
+               and self.rec(expr.variables, other.variables, *args, **kwargs) \
+               and self.rec(expr.values, other.values, *args, **kwargs)
+
+    def map_slice(self, expr, other, *args, **kwargs):
+        return self._map_multichild_expr(expr, other, *args, **kwargs)
```

Comment:
I could imagine two fast paths in `self.rec`, to parallel what `__eq__` does currently:

- Return `True` if `self is other`.
- Return `False` if the hashes differ.